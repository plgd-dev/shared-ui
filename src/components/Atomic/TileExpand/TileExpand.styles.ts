import { css } from '@emotion/react'

import { ThemeType, get } from '../_theme'
import { fontPrimary, fontSecondary } from '../_utils/commonStyles'

export const tileExpand = (theme: ThemeType) => css`
    border-radius: 8px;
    border: 1px solid ${get(theme, 'TileExpand.border')};
    overflow: hidden;
`

export const header = (theme: ThemeType) => css`
    padding: 16px 24px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    cursor: pointer;
    background: ${get(theme, 'TileExpand.header.background')};

    &:hover {
        background: ${get(theme, 'TileExpand.header.hover.background')};

        .expander {
            color: ${get(theme, 'TileExpand.header.hover.expander.color')};
        }
    }
`

export const left = css``

export const tagSpace = css`
    padding-left: 6px;
`

const headlineBase = css`
    font: ${fontSecondary};
    font-size: 16px;
    font-style: normal;
    font-weight: 700;
    line-height: 150%;
    letter-spacing: -0.5px;
`

export const title = (theme: ThemeType) => css`
    ${headlineBase};
    color: ${get(theme, 'TileExpand.title.color')};
    display: flex;
    align-items: center;
`

export const time = (theme: ThemeType) => css`
    font: ${fontPrimary};
    font-size: 14px;
    font-style: normal;
    font-weight: 400;
    line-height: 22px;
    color: ${get(theme, 'TileExpand.time.color')};
    padding-top: 2px;
`

export const right = css`
    flex: 0 0 32px;
`

export const expander = (theme: ThemeType) => css`
    color: ${get(theme, 'TileExpand.expander.color')};
    background: ${get(theme, 'TileExpand.expander.background')};
    padding: 6px;
    border-radius: 8px;
    transition: all 0.3s;
    width: 32px;
    height: 32px;

    &:hover {
        color: ${get(theme, 'TileExpand.expander.hover.color')};
    }
`

export const content = (theme: ThemeType) => css`
    padding: 0 24px 24px 24px;
    border-top: 1px solid ${get(theme, 'TileExpand.border')};
`

export const group = css`
    padding-top: 16px;
`

export const groupHeadline = (theme: ThemeType) => css`
    ${headlineBase};
    color: ${get(theme, 'TileExpand.title.color')};
    display: flex;
    align-items: center;
    padding-bottom: 16px;
`

export const error = (theme: ThemeType) => css`
    padding: 12px;
    border-radius: 8px;
    background: ${get(theme, 'TileExpand.error.background')};
`

export const errorTitle = (theme: ThemeType) => css`
    font-family: ${fontPrimary};
    font-size: 14px;
    font-style: normal;
    font-weight: 600;
    line-height: 22px;
    color: ${get(theme, 'TileExpand.error.title.color')};
    margin-bottom: 6px;
`

export const errorMessage = (theme: ThemeType) => css`
    font-family: ${fontPrimary};
    font-size: 12px;
    font-style: normal;
    font-weight: 400;
    line-height: 18px;
    color: ${get(theme, 'TileExpand.error.message.color')};
`

export const infoLine = (theme: ThemeType) => css`
    padding: 16px;
    border-radius: 8px;
    background: ${get(theme, 'TileExpand.info.background')};
    font-family: ${fontPrimary};
    font-size: 14px;
    font-style: normal;
    font-weight: 400;
    line-height: 22px;

    & + & {
        margin-top: 4px;
    }
`

export const attribute = (theme: ThemeType) => css`
    color: ${get(theme, 'TileExpand.info.attribute.color')};
`

export const value = (theme: ThemeType) => css`
    display: flex;
    align-items: center;
    color: ${get(theme, 'TileExpand.info.value.color')};
    justify-content: flex-end;
`

export const valueRaw = (theme: ThemeType) => css`
    color: ${get(theme, 'TileExpand.info.value.color')};
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    width: 100%;
    text-align: right;
`

export const icon = (theme: ThemeType) => css`
    margin-left: 6px;
    color: ${get(theme, 'TileExpand.info.value.icon.color')};
    cursor: pointer;

    &:hover {
        color: ${get(theme, 'TileExpand.info.value.icon.hover.color')};
    }
`
