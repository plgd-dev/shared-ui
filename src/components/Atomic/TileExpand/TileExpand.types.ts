import { ReactNode } from 'react'

export type Information = {
    attribute: string
    certFormat?: boolean
    copyValue?: string
    value: string
}

export type Props = {
    className?: string
    defaultOpen?: boolean
    error?: {
        groupTitle: string
        title: string
        message: string
    }
    hasExpand?: boolean
    i18n: {
        copy: string
    }
    information?: {
        groupTitle: string
        rows: Information[]
    }
    title: string
    statusTag?: ReactNode
    time?: ReactNode
}

export const defaultProps: Partial<Props> = {
    hasExpand: true,
}
