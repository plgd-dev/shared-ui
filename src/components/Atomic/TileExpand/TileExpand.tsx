import React, { forwardRef, useState } from 'react'
import { AnimatePresence, motion } from 'framer-motion'

import { Information, Props, defaultProps } from './TileExpand.types'
import * as styles from './TileExpand.styles'
import IconArrowDownNoPadding from '../Icon/components/IconArrowDownNoPadding'
import { convertSize, IconCopy } from '../Icon'
import Row from '../Grid/Row'
import Column from '../Grid/Column'
import Tooltip from '../Tooltip'
import { copyToClipboard } from '../../../common/utils'

const TileExpand = forwardRef<HTMLDivElement, Props>((props, ref) => {
    const { className, defaultOpen, error, hasExpand, i18n, information, statusTag, title, time } = { ...defaultProps, ...props }

    const [show, setShow] = useState(defaultOpen ?? false)

    return (
        <div className={className} css={styles.tileExpand} ref={ref}>
            <div css={styles.header} onClick={hasExpand ? () => setShow(!show) : undefined}>
                <div css={styles.left}>
                    <div css={styles.title}>
                        {title}
                        {!!statusTag && <span css={styles.tagSpace}>{statusTag}</span>}
                    </div>
                    {!!time && <div css={styles.time}>{time}</div>}
                </div>
                {hasExpand && (
                    <div css={styles.right}>
                        <motion.div
                            animate={{
                                rotate: show ? 180 : 0,
                            }}
                            className='expander'
                            css={styles.expander}
                            transition={{
                                duration: 0.3,
                            }}
                        >
                            <IconArrowDownNoPadding {...convertSize(20)} />
                        </motion.div>
                    </div>
                )}
            </div>
            <AnimatePresence>
                {show && (
                    <motion.div
                        animate={{ opacity: 1, height: 'auto' }}
                        exit={{
                            opacity: 0,
                            height: 0,
                        }}
                        initial={{ opacity: 0, height: 0 }}
                        transition={{
                            duration: 0.3,
                        }}
                    >
                        <div css={styles.content}>
                            {!!error && (
                                <div css={styles.group}>
                                    <div css={styles.groupHeadline}>{error.groupTitle}</div>
                                    <div css={styles.error}>
                                        <div css={styles.errorTitle}>{error.title}</div>
                                        <div css={styles.errorMessage}>{error.message}</div>
                                    </div>
                                </div>
                            )}
                            {!!information && (
                                <div css={styles.group}>
                                    <div css={styles.groupHeadline}>{information.groupTitle}</div>
                                    {information.rows.map((info: Information, k) => (
                                        <div css={styles.infoLine} key={info.attribute || k}>
                                            <Row>
                                                <Column size={5}>
                                                    <div css={styles.attribute}>{info.attribute}</div>
                                                </Column>
                                                <Column size={7}>
                                                    <div css={styles.value}>
                                                        <span css={styles.valueRaw}>{info.value}</span>
                                                        <Tooltip
                                                            content={i18n.copy}
                                                            css={styles.icon}
                                                            id={`tooltip-group-${info.attribute}`}
                                                            portalTarget={undefined}
                                                        >
                                                            <IconCopy
                                                                onClick={() => copyToClipboard(info?.copyValue || (info.value as string), info.certFormat)}
                                                            />
                                                        </Tooltip>
                                                    </div>
                                                </Column>
                                            </Row>
                                        </div>
                                    ))}
                                </div>
                            )}
                        </div>
                    </motion.div>
                )}
            </AnimatePresence>
        </div>
    )
})

TileExpand.displayName = 'TileExpand'

export default TileExpand
