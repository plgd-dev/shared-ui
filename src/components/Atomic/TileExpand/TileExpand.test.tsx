import { render } from '../../../../test/jest-wrapper'
import TileExpand from './TileExpand'

describe('<TileExpand>', () => {
    it('renders title and status tag when provided', () => {
        const { getByText, asFragment } = render(<TileExpand i18n={{ copy: 'copy' }} statusTag='Test Status' title='Test Title' />)

        expect(getByText('Test Title')).toBeInTheDocument()
        expect(getByText('Test Status')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders time when provided', () => {
        const { getByText } = render(<TileExpand i18n={{ copy: 'copy' }} time='Test Time' title='Title' />)
        expect(getByText('Test Time')).toBeInTheDocument()
    })
})
