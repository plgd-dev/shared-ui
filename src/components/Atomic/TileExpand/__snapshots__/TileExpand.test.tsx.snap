// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TileExpand> renders title and status tag when provided 1`] = `
<DocumentFragment>
  .emotion-0 {
  border-radius: 8px;
  border: 1px solid;
  overflow: hidden;
}

.emotion-1 {
  padding: 16px 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  cursor: pointer;
}

.emotion-3 {
  font: 'Circular Pro',sans-serif;
  font-size: 16px;
  font-style: normal;
  font-weight: 700;
  line-height: 150%;
  letter-spacing: -0.5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  padding-left: 6px;
}

.emotion-5 {
  -webkit-flex: 0 0 32px;
  -ms-flex: 0 0 32px;
  flex: 0 0 32px;
}

.emotion-6 {
  padding: 6px;
  border-radius: 8px;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 32px;
  height: 32px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          Test Title
          <span
            class="emotion-4"
          >
            Test Status
          </span>
        </div>
      </div>
      <div
        class="emotion-5"
      >
        <div
          class="expander emotion-6"
        >
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.59 6.912a.833.833 0 1 0-1.18 1.179l5 5a.833.833 0 0 0 1.18 0l5-5a.833.833 0 0 0-1.18-1.179L10 11.322z"
              fill="currentcolor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
